hbase-server/src/test/java/org/apache/hadoop/hbase/replication/regionserver/TestMetaRegionReplicaReplicationEndpoint.java

```diff
@@ -225,6 +227,38 @@ public void testCatalogReplicaReplicationWithFlushAndCompaction() throws Excepti
     }
   }
 
+  @Test
+  public void testCatalogReplicaReplicationWALRolledAndDeleted() throws Exception {
+    Connection connection = ConnectionFactory.createConnection(HTU.getConfiguration());
+    TableName tableName = TableName.valueOf("hbase:meta");
+    Table table = connection.getTable(tableName);
+    try {
+      MiniHBaseCluster cluster = HTU.getHBaseCluster();
+      HRegionServer hrs = cluster.getRegionServer(cluster.getServerHoldingMeta());
+      ReplicationSource source = (ReplicationSource) hrs.getReplicationSourceService()
+        .getReplicationManager().catalogReplicationSource.get();
+      ((ReplicationPeerImpl) source.replicationPeer).setPeerState(false);
```

Comment:
Is this enough to make sure that the replication is already stopped? The check is in another thread, we'd better add some checks here to confirm that the replication source is stopped.

Reply:
Ok, there's a small chance for race conditions here, although I have run this many times and never faced it. 

Nevertheless, I'm putting extra checks to make sure we just proceed with adding data once we are sure the source wal reader knows the peer is disabled now. Please check, @Apache9 .